homeassistant/components/sharkiq/vacuum.py

```diff
@@ -0,0 +1,286 @@
+"""Shark IQ Wrapper."""
+
+
+import logging
+from typing import Dict, Iterable, Optional
+
+from sharkiqpy import OperatingModes, PowerModes, Properties, SharkIqVacuum
+
+from homeassistant.components.vacuum import (
+    STATE_CLEANING,
+    STATE_DOCKED,
+    STATE_IDLE,
+    STATE_PAUSED,
+    STATE_RETURNING,
+    SUPPORT_BATTERY,
+    SUPPORT_FAN_SPEED,
+    SUPPORT_LOCATE,
+    SUPPORT_PAUSE,
+    SUPPORT_RETURN_HOME,
+    SUPPORT_START,
+    SUPPORT_STATE,
+    SUPPORT_STATUS,
+    SUPPORT_STOP,
+    StateVacuumEntity,
+)
+
+from .const import DOMAIN, SHARK
+from .update_coordinator import SharkIqUpdateCoordinator
+
+LOGGER = logging.getLogger(__name__)
+
+# Supported features
+SUPPORT_SHARKIQ = (
+    SUPPORT_BATTERY
+    | SUPPORT_FAN_SPEED
+    | SUPPORT_PAUSE
+    | SUPPORT_RETURN_HOME
+    | SUPPORT_START
+    | SUPPORT_STATE
+    | SUPPORT_STATUS
+    | SUPPORT_STOP
+    | SUPPORT_LOCATE
+)
+
+OPERATING_STATE_MAP = {
+    OperatingModes.PAUSE: STATE_PAUSED,
+    OperatingModes.START: STATE_CLEANING,
+    OperatingModes.STOP: STATE_IDLE,
+    OperatingModes.RETURN: STATE_RETURNING,
+}
+
+FAN_SPEEDS_MAP = {
+    "Eco": PowerModes.ECO,
+    "Normal": PowerModes.NORMAL,
+    "Max": PowerModes.MAX,
+}
+
+STATE_RECHARGING_TO_RESUME = "recharging_to_resume"
+
+# Attributes to expose
+ATTR_ERROR_CODE = "last_error_code"
+ATTR_ERROR_MSG = "last_error_message"
+ATTR_LOW_LIGHT = "low_light"
+ATTR_RECHARGE_RESUME = "recharge_and_resume"
+ATTR_RSSI = "rssi"
+
+
+class SharkVacuumEntity(StateVacuumEntity):
+    """Shark IQ vacuum entity."""
+
+    def __init__(self, sharkiq: SharkIqVacuum, coordinator: SharkIqUpdateCoordinator):
+        """Create a new SharkVacuumEntity."""
+        if sharkiq.serial_number not in coordinator.shark_vacs:
+            raise RuntimeError(
```

Comment:
When can this happen?

Reply:
It should never happen.  This is just being over-cautious.  Happy to remove or comment.